app/client/ts/video-cache.ts

```diff
@@ -0,0 +1,125 @@
+import Video from "./video-model";
+import Config from "./global-config";
+import * as stream from "stream";
+import { Maybe } from "./maybe";
+import { safeClient } from "./util";
+import { SafeFile, DnsHomeDirectory, NfsFileData } from "safe-launcher-client";
+
+const CONFIG: Config = Config.getInstance();
+
+interface Cache {
+    [xorName: string]: {
+        video: Video,
+        accessTime: Date
+    }
+}
+
+type CacheStateStringy = {
+    xorName: string;
+    name: string;
+}
+
+export default class VideoCache {
+    private static readonly CACHE_STATE_FILENAME = 'cachestate';
+    private static INSTANCES: {
+        [longName: string]: VideoCache
+    } = {};
+
+    private cache: Cache;
+    private numEntries: number;
+
+    private constructor(private longName: string, oldState: Maybe<CacheStateStringy>) {
+        oldState.caseOf({
+            nothing: () => {
+                this.cache = {};
+                this.numEntries = 0;
+            },
+            just: (state: CacheStateStringy) => {
+                this.cache = {};
+                this.numEntries = Infinity;
+            }
+        });
+    }
+
+    static async getInstance(longName: string) {
+        if (VideoCache.INSTANCES[longName] == null) {
+            VideoCache.INSTANCES[longName] = new VideoCache(longName, await VideoCache.readFromDisk(longName));
+        }
+        return VideoCache.INSTANCES[longName];
+    }
+
+    private garbageCollect(): void {
+        // sort by oldest first
+        const oldKeys: string[] = Object.keys(this.cache).sort((key1: string, key2: string) => {
+            if (this.cache[key1].accessTime < this.cache[key2].accessTime) return -1;
+            else if (this.cache[key1].accessTime > this.cache[key2].accessTime) return 1;
+            else return 0;
+        });
+
+        // delete half of the oldest cache entries
+        for (let i = 0; i < oldKeys.length / 2; i++) {
+            delete this.cache[oldKeys[i]];
+        }
+        this.numEntries = Object.keys(this.cache).length;
+    }
+
+    async getFromXorName(xorName: string): Promise<Video> {
+        if (this.cache[xorName] == null) {
+            if (this.numEntries + 1 > CONFIG.MAX_CACHE_SIZE) {
+                this.garbageCollect();
+            }
+
+            this.cache[xorName] = {
+                video: await Video.readFromStringXorName(xorName),
+                accessTime: new Date()
+            };
+            this.numEntries += 1;
+        } else {
+            this.cache[xorName].accessTime = new Date();
+        }
+
+        console.log('cache state:', this.cache);
+        return this.cache[xorName].video;
+    }
+
+    stateToString(): string {
+        //const ret = Object.keys(this.cache).map((xorName: string) => return {
+        return JSON.stringify({
+            xorName: '',
+            name: ''
+        });
+    }
+
+    private static async readFromDisk(longName): Promise<Maybe<CacheStateStringy>> {
```

Comment:
What does Nothing represent that is different than an empty CacheStateStringy object? A comment explaining might be a little nice. (non merge blocking).